### ISSUE TITLE
bindings: python: investigate free threaded builds

### ISSUE THREAD

**Issue by vfazio** [brgl/libgpiod#117]
> 
> Ha! Is there any writeup on this? I would definitely want to make sure the bindings to work fine without the GIL.

https://peps.python.org/pep-0703/

https://docs.python.org/3/whatsnew/3.13.html#free-threaded-cpython

https://docs.python.org/3/c-api/module.html#c.Py_mod_gil

https://py-free-threading.github.io/porting/

_Originally posted by @vfazio in https://github.com/brgl/libgpiod/issues/93#issuecomment-2405210680_

**Comment by vfazio** [comment]
I don't know if we should be defensive and throw a compile error if a gil-less build is attempted until we've had a chance to test and validate everything works as expected.

**Comment by brgl** [comment]
Does any distro ship GIL-less python 3.13 yet? I'm pretty sure it will make its way into yocto soon, this is when I'll most likely start playing with it. Until then, I don't think we should really do anything about it.

**Comment by vfazio** [comment]
@brgl looks like Fedora ships 3.13 as part of f41+: https://docs.fedoraproject.org/en-US/fedora/latest/release-notes/developers/

and it looks like there are build changes and rpms for freethreading, at least for f42, but they're probably also available for f41

https://src.fedoraproject.org/rpms/python3.13/c/84d78b385a57306af12b04922f60d516e6e8be49

https://koji.fedoraproject.org/koji/buildinfo?buildID=2599772

https://koji.fedoraproject.org/koji/rpminfo?rpmID=40917017

**Comment by vfazio** [comment]
@brgl if you're really bored, do you want to try testing out https://github.com/vfazio/libgpiod/tree/vfazio-upstream-proposed ?

I've added beta free-threaded support.

I've run the tests under normal python and under a free-threaded build and have not seen any issues when appropriate precautions are taken.

**Comment by vfazio** [comment]
I'm not proud of it, but I've rebased the branch and gotten pretty consistent passing results in my numerous tests.

https://github.com/vfazio/libgpiod/pull/11

As I can't run this on my laptop, I'm at the whims of Github's CI runners. I'll try to do some intensive runs of the patches at work... maybe 50k+ iterations or so.

I need to add detail to the commits and add documentation to the readme, which is apparently out of date and will be even more out of date when the Meson patches get applied since the readme advises running `make python-tests-run`

**Comment by vfazio** [comment]
ok well, i forgot i can remote into work so i just ran 500k iterations

```
vfazio@vfazio4:~/development/libgpiod-freethread/bindings/python$ git diff
diff --git a/bindings/python/tests/tests_threading.py b/bindings/python/tests/tests_threading.py
index bd10129..063dd5c 100644
--- a/bindings/python/tests/tests_threading.py
+++ b/bindings/python/tests/tests_threading.py
@@ -105,8 +105,8 @@ class ThreadedTestCase(TestCase):
     def setUpClass(cls) -> None:
         cls.NUM_THREADS = 4
         # we want to stress test free threaded builds a bit more
-        cls.ITERATIONS = 200 if is_free_threaded() else 20
-        cls.TIMEOUT = 2
+        cls.ITERATIONS = 500_000 if is_free_threaded() else 20
+        cls.TIMEOUT = 60


 class Chip(ThreadedTestCase):
```

```
root@vfazio4:/home/vfazio/development/libgpiod-freethread/bindings/python# LD_PRELOAD=`gcc -print-file-name=libasan.so` make python-tests-run
PYTHONPATH=/home/vfazio/development/libgpiod-freethread/bindings/python \
LD_LIBRARY_PATH=/home/vfazio/development/libgpiod-freethread/lib/.libs/:\
/home/vfazio/development/libgpiod-freethread/tests/gpiosim/.libs/ \
/home/vfazio/development/libgpiod-freethread/bindings/python/.venv/bin/python -B -m tests
............................................DEBUG:tests.tests_threading:3 local_count=564133
DEBUG:tests.tests_threading:2 local_count=435777
............DEBUG:tests.tests_threading:3 local_count=559529
DEBUG:tests.tests_threading:2 local_count=567603
..........................................................................................................
----------------------------------------------------------------------
Ran 162 tests in 173.828s

OK
```

**Comment by brgl** [comment]
Does this mean, we're good with free-threaded mode? Awesome!

**Comment by vfazio** [comment]
> Does this mean, we're good with free-threaded mode? Awesome!

Yes, once i can find the time to polish up the commits so people don't scoff at my poor coding

**Comment by vfazio** [comment]
I'll be off work next week so may have some time to wrap this up finally so we can cut a new version.

**Comment by vfazio** [comment]
enabled as part of series: https://lore.kernel.org/linux-gpio/20260522200419.105496-1-vfazio@gmail.com/

**Comment by vfazio** [comment]
@brgl wanted to chat here as it's (selfishly) easier for me...

re:

> Do we need to document which classes and methods need synchronization?

What are you envisioning? Something in the API.rst, README.md or something in the class and function docstrings? I'm also not sure how detailed you want me to be so just lmk what you're looking for so I can address it.

**Comment by brgl** [comment]
> > Do we need to document which classes and methods need synchronization?
> 
> What are you envisioning? Something in the API.rst, README.md or something in the class and function docstrings? I'm also not sure how detailed you want me to be so just lmk what you're looking for so I can address it.

I was thinking about something like what you have in tests_threading.py already in pydocs: for each class just add a line: `Synchronization: none required` or `Synchronization: objects of this class must be protected from concurrent access`?

**Comment by vfazio** [comment]
Merged
